packages/pigeon/lib/ast.dart

```diff
@@ -82,35 +74,101 @@ class Api extends Node {
   }
 }
 
+/// A parameter to a generic entity.  For example, "String" to "List<String>".
+class TypeArgument {
+  /// Constructor for [TypeArgument].
+  TypeArgument({
+    required this.dataType,
+    required this.isNullable,
+    this.typeArguments,
+  });
+
+  /// A string representation of the base datatype.
+  final String dataType;
+
+  /// The type arguments to this [TypeArgument].
+  final List<TypeArgument>? typeArguments;
+
+  /// True if the type is nullable.
+  final bool isNullable;
+
+  @override
+  String toString() {
+    return '(TypeArgument dataType:$dataType isNullable:$isNullable typeArguments:$typeArguments)';
+  }
+}
+
+/// An entity that represents a typed concept, like a [TypeArgument] or [Field].
+abstract class TypedEntity {
+  /// The data-type of the entity (ex 'String' or 'int').
+  String get dataType;
+
+  /// The type arguments to the entity.
+  List<TypeDeclaration>? get typeArguments;
+
+  /// True if the type is nullable.
+  bool get isNullable;
+}
+
+/// Represents a type declaration.
+class TypeDeclaration implements TypedEntity {
+  /// Constructor for [TypeDeclaration].
+  TypeDeclaration({
+    required this.dataType,
+    required this.isNullable,
+    this.typeArguments,
+  });
+
+  @override
+  final String dataType;
+
+  @override
+  final List<TypeDeclaration>? typeArguments;
+
+  @override
+  final bool isNullable;
+
+  @override
+  String toString() {
+    return '(TypeDeclaration dataType:$dataType isNullable:$isNullable typeArguments:$typeArguments)';
+  }
+}
+
 /// Represents a field on a [Class].
-class Field extends Node {
+class Field extends Node implements TypedEntity {
   /// Parametric constructor for [Field].
   Field({
     required this.name,
-    required this.dataType,
-    required this.isNullable,
-    this.typeArguments,
+    required String dataType,
+    required bool isNullable,
+    List<TypeDeclaration>? typeArguments,
     this.offset,
-  });
+  }) : type = TypeDeclaration(
+            dataType: dataType,
+            isNullable: isNullable,
+            typeArguments: typeArguments);
 
   /// The name of the field.
   String name;
 
-  /// The data-type of the field (ex 'String' or 'int').
-  String dataType;
-
   /// The offset in the source file where the field appears.
   int? offset;
 
-  /// True if the datatype is nullable (ex `int?`).
-  bool isNullable;
+  /// The type of the [Field].
+  TypeDeclaration type;
 
-  /// Type parameters used for generics.
-  List<Field>? typeArguments;
+  @override
+  String get dataType => type.dataType;
+
+  @override
+  bool get isNullable => type.isNullable;
+
+  @override
+  List<TypeDeclaration>? get typeArguments => type.typeArguments;
 
   @override
   String toString() {
-    return '(Field name:$name dataType:$dataType)';
+    return '(Field name:$name dataType:$dataType typeArguments:$typeArguments)';
```

Comment:
Note that you still have the bug here that will be fixed by replacing the fields here with `TypeDeclaration`.